Improve container and imagefs stats

xinfengliu opened the pull request:
We have seen high CPU usage in `dockerd` and `cri-dockerd` in some cases. The profiling data shows high CPU was caused by calculating container RW layer size.  Also there's an issue report in https://github.com/Mirantis/cri-dockerd/issues/135

This PR does:

- Limit the concurrency of `ListContainerStats`.
  Currently the number of concurrency is number of k8s containers, this could cause thunderstorm loads on big nodes running a large number of pods. This PR uses [pipeline pattern](https://pkg.go.dev/golang.org/x/sync/errgroup#example-Group-Pipeline) ~to use at most half of number of CPUs for `ListContainerStats`~.  Please see later comments in this PR for updates.

- Cache container stats with dynamic TTL based on the time taken for getting stats.
   Use another cache implementation to overcome the below shortcomings of `k8s.io/client-go/tools/cache`: 
   1. it could not set different cache TTL for different entries
   2. it could cause time-consuming operations to fire in parallel. E.g. the `InspectContainerWithSize` could take minutes to finish, if multiple requests come in, it will cause multiple `InspectContainerWithSize` to run, thus adding high load to the system. 
   - This customized cache prevents time-consuming operations from running in parallel. In addition, it adds `lastTimeTaken` to the cache TTL to make cache TTL dynamic. Normally the operation takes some ms or a few seconds, adding `lastTimeTaken` does not affect much to TTL. Some resource consuming operations may take minutes to finish, it does not make sense to keep the low TTL since the stats are already not realtime, make TTL longer helps reduce the load.

- cache imageFS stats

- For linux, calculate imageFS using docker root directory (obsoletes https://github.com/Mirantis/cri-dockerd/pull/221). [ImageFs](https://kubernetes.io/docs/concepts/scheduling-eviction/node-pressure-eviction/#eviction-signals) means the filesystem that container runtimes use to store container images and container writable layers. Only counting `image` directory itself is wrong and does not help `kubelet` make correct decisions for garbage collections or evictions.

- Rename `image_linux.go` and `image_windows.go` to `imagefs_linux.go` and `imagefs_windows.go` to better reflect the file purposes.

nwneisen commented on core/stats.go:
```diff
@@ -57,24 +69,60 @@ func (ds *dockerService) ListContainerStats(
-				mtx.Unlock()
-			} else if err != nil {
-				logrus.Error(err, " Failed to get stats from container "+container.Id)
+	g, ctx := errgroup.WithContext(ctx)
```
Can I have you add some inline comments to make it easier to remember what this code is doing the next time around

xinfengliu replied:
I just added some comments in a new commit, please let me know if it's OK.

xinfengliu commented on core/stats.go:
```diff
@@ -57,24 +69,60 @@ func (ds *dockerService) ListContainerStats(
+		for _, container := range listResp.Containers {
+			select {
+			case containerIDCh <- container.Id:
+			case <-ctx.Done():
```
`case <-ctx.Done()` here is used for return early after the context is cancelled. It means a worker failed or the client has cancelled the (parent) context. 

Successfully iterating all the containers here just means we finish sending the jobs to the job queue `containerIDCh`, the jobs are not finished yet, thus it will not receive anything from `<-ctx.Done()` unless an error happens.

nwneisen reviewed (approved):
LGTM! Thank you!

neersighted commented on core/stats.go:
```diff
@@ -57,24 +69,67 @@ func (ds *dockerService) ListContainerStats(
+
+	// limit concurrency to half of number of CPU
+	statsCh := make(chan *runtimeapi.ContainerStats)
+	numWorkers := runtime.NumCPU() / 2
+	if numWorkers == 0 {
+		numWorkers = 1
+	}
```
```suggestion
  // we must have at least one worker
	numWorkers := math.Max(runtime.NumCPU() / 2, 1)
```

xinfengliu replied:
Hi @neersighted , using `math.Max` needs to convert between integer and float64, it's boring. And I did not want to import a new package for just `max()`. In golang 1.21, there's a built-in function `max`. Shall we consider it?

nwneisen replied:
FWIW I also thought about suggesting this change but decided the original code kept things simple, didn't bring in other libs, and wont have any surprise behavior down the road :man_shrugging:

neersighted replied:
Ah, that's fair. I'm fine with requiring go 1.21+, but cri-dockerd is compiled as part of MCR and therefore is subject to the limitations of the compiler toolchain.

cc @corhere WRT whether or not a `go 1.21` line in `go.mod` would be problematic.

corhere replied:
Requiring Go 1.21 would be problematic for MCR. I'm still working on migrating us to Go 1.20!

neersighted commented on core/docker_service.go:
```diff
@@ -261,6 +261,16 @@ func NewDockerService(
+	// Do housekeeping every minute to clean up cache entries that are longer than 1 hour.
+	go func() {
+		for {
+			ContainerStatsCache.ClearByAge(time.Hour)
```
How expensive can this call be? If it's not expensive, the following feels more correct to me:
```go
for range time.Tick(time.Minute) {
    ContainerStatsCache.ClearByAge(time.Hour)
}
```

xinfengliu replied:
Not expensive, just iterating a map with at most hundreds of pointers. 
Updated per your suggestion in the new commit.

xinfengliu commented:
I pushed a few more commits over the weekend based on my testing and debugging.

- Use `statfs` for `imageFsInfo()` on Linux
   I think the `usedBytes` and `iNodesUsed` are only meaningful at the filesystem level for `kubelet`'s decision on garbage collection or eviction.

- Reduce unnecessary function calls
   - `ListContainerStats` calls `ds.ListContainers` just for getting a list of container IDs. Changed to use a lightweight version.
   - `getContainerStats` calls `ds.ContainerStatus` which calls again `InspectContainer`, at earlier time `InspectContainerWithSize` has been called. Changed to remove the call to `InspectContainer`.
   
- Refactor the cache for `docker info` and `docker version`
   When `docker info` takes long time, it could cause multiple incoming requests (e.g.  `getContainerStats`) for getting `docker info` from the `k8s.io/client-go/tools/cache` fire multiple `docker info` request to `dockerd` when they all see the value is not in the cache. I used a version with debugging info to capture this: 
```
Sep 10 12:25:09 ub2004-3 cri-dockerd-mke[625387]: time="2023-09-10T12:25:09+08:00" level=info msg="Calling dockerd for docker info"
Sep 10 12:25:09 ub2004-3 cri-dockerd-mke[625387]: time="2023-09-10T12:25:09+08:00" level=info msg="Calling dockerd for docker info"
Sep 10 12:25:09 ub2004-3 cri-dockerd-mke[625387]: time="2023-09-10T12:25:09+08:00" level=info msg="Calling dockerd for docker info"
Sep 10 12:25:09 ub2004-3 cri-dockerd-mke[625387]: time="2023-09-10T12:25:09+08:00" level=info msg="Calling dockerd for docker info"
Sep 10 12:25:09 ub2004-3 cri-dockerd-mke[625387]: time="2023-09-10T12:25:09+08:00" level=info msg="Calling dockerd for docker info"
Sep 10 12:25:09 ub2004-3 cri-dockerd-mke[625387]: time="2023-09-10T12:25:09+08:00" level=info msg="Calling dockerd for docker info"
Sep 10 12:25:09 ub2004-3 cri-dockerd-mke[625387]: time="2023-09-10T12:25:09+08:00" level=info msg="Calling dockerd for docker info"
Sep 10 12:25:09 ub2004-3 cri-dockerd-mke[625387]: time="2023-09-10T12:25:09+08:00" level=info msg="Calling dockerd for docker info"
...
```
I changed to use the home-made cache.

- Rework the concurrency controller for `ListContainerStats`
  Although `getContainerStats` calls `ContainerStatsOneShot`, this could still cause the response time has about `1s` delay,  we can check it by `time curl --unix-socket /var/run/docker.sock "http:/api/containers/<a-running-container-name>/stats?oneshot=1&stream=0"`. Therefore, we want to set a bigger value for the number of workers to avoid too long delays. Consider a common node with 8 CPU running dozens of pods, `NumCPU() * 6` may be a moderate number.

neersighted commented on core/imagefs_linux.go:
```diff
@@ -20,31 +20,31 @@ limitations under the License.
 		return nil, err
 	}
 
-	bytes, inodes, err := dirSize(filepath.Join(info.DockerRootDir, "image"))
```
This additional work looks promising to me (need to spend more time looking at specifics later in the week when I find time); it's worth noting that this will conflict with/obsolete https://github.com/Mirantis/cri-dockerd/pull/221.

cc @surik

xinfengliu commented:
Sorry for the messy commits, I just squashed them. 

Highlighted recent changes:

- Pick up https://github.com/kubernetes/kubernetes/pull/104287 which `cri-dockred` had missed. 
  This also means we use a new `dockerRootDir` field in `DockerService`, this eliminates many unnecessary `docker info` calls.

- I created https://github.com/moby/moby/pull/46448 for addressing `dockerd` stats collector scalability issue.

neersighted reviewed (commented):
Looks like `store.NewObjectCache` is unused with these changes -- can we please add removing it to this PR?

xinfengliu commented:
> Looks like `store.NewObjectCache` is unused with these changes -- can we please add removing it to this PR?

Done. Thanks for pointing it out.

xinfengliu commented:
@neersighted 
Could you help get https://github.com/moby/moby/pull/46448 reviewed? Thanks. (I don't have the permission to request reviewers.)

xinfengliu commented:
Some new updates:

- Since the time for calculating container RW layer size is unpredictable (In some cases, it could be multiple minutes, which exceeds the default docker client timeout 1m59s), it's beneficial to decouple the container RW layer size calculation.

- The ImageFS used_bytes is now computed via docker images API, which is consistent with the way used by `cadvisor`, although I still think it's not meaningful since docker volumes, builder caches could also consume a lot of spaces.

New commits have been rebased with master branch and squashed.

nwneisen commented on core/stats_windows.go:
```diff
@@ -71,27 +66,22 @@ func (ds *dockerService) getContainerStats(containerID string) (*runtimeapi.Cont
-		return nil, err
-	}
-	status := statusResp.GetStatus()
+	// used bytes from image are not implemented on Windows
```
Can we remove this commented code?

xinfengliu replied:
Will do when I'm back from holidays.

xinfengliu commented on core/docker_service.go:
```diff
@@ -66,11 +67,9 @@ const (
-	// The expiration time of 'docker info' cache.
-	infoCacheTTL = 60 * time.Second
-	maxMsgSize   = 1024 * 1024 * 16
+	systemInfoCacheMinTTL = time.Minute
```
`systemInfoCacheMinTTL` is used for both `docker version` and `docker info` cache, so I wanted to use a const to represent it.

xinfengliu commented on core/docker_service.go:
```diff
@@ -216,23 +216,18 @@ func NewDockerService(
+	if err != nil {
+		return nil, fmt.Errorf("Failed to execute Info() call to the Docker client")
+	}
+	logrus.Infof("Docker Info: %+v", dockerInfo)
```
I remember it's also in previous versions. We can change it to debugging log if you feel OK.

nwneisen replied:
I'm not worried if it was there previously. The information does seem better as a debugging log to me though.

xinfengliu commented on core/imagefs_linux.go:
```diff
@@ -0,0 +1,82 @@
+			totalImageSize += uint64(i.Size)
+		} else { // docker version >= 23.0
+			totalImageSize += (uint64(i.Size) - uint64(i.SharedSize))
+			sharedSizeMap[i.SharedSize] = struct{}{}
```
Yes, you are right. However, this is all I currently can do because I could not find the image layer chain info in the data returned from `docker images`.  This way is still better than what cadvisor stats does: cadvisor does not de-dup the shared size (at least in kube 1.24.6).

nwneisen replied:
Sounds good. I just wanted to understand our current limitations.

xinfengliu commented:
@nwneisen I pushed a new commit per your recent feedbacks. Please let me know if there's anything else I need to do.

nwneisen commented:
@xinfengliu The changes look good. I will go ahead an merge this in. Thank you for the commit

zhangguanzhang commented:
after update to `v0.3.5`, fixed this:
```
ERRO[0037] Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running? Failed to get stats from container ...
ERRO[0037] Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running? Failed to get stats from container ...
ERRO[0037] Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running? Failed to get stats from container ...
```
